src/backend/linalg_internal_cpu/Mul_internal.hpp

```diff
@@ -1,755 +1,70 @@
-    void Mul_internal_btu64(boost::intrusive_ptr<Storage_base> &out,
-                            boost::intrusive_ptr<Storage_base> &Lin,
-                            boost::intrusive_ptr<Storage_base> &Rin, const unsigned long long &len,
-                            const std::vector<cytnx_uint64> &shape,
-                            const std::vector<cytnx_uint64> &invmapper_L,
-                            const std::vector<cytnx_uint64> &invmapper_R);
-    void Mul_internal_bti32(boost::intrusive_ptr<Storage_base> &out,
-                            boost::intrusive_ptr<Storage_base> &Lin,
-                            boost::intrusive_ptr<Storage_base> &Rin, const unsigned long long &len,
-                            const std::vector<cytnx_uint64> &shape,
-                            const std::vector<cytnx_uint64> &invmapper_L,
-                            const std::vector<cytnx_uint64> &invmapper_R);
-    void Mul_internal_btu32(boost::intrusive_ptr<Storage_base> &out,
-                            boost::intrusive_ptr<Storage_base> &Lin,
-                            boost::intrusive_ptr<Storage_base> &Rin, const unsigned long long &len,
-                            const std::vector<cytnx_uint64> &shape,
-                            const std::vector<cytnx_uint64> &invmapper_L,
-                            const std::vector<cytnx_uint64> &invmapper_R);
-    void Mul_internal_btu16(boost::intrusive_ptr<Storage_base> &out,
-                            boost::intrusive_ptr<Storage_base> &Lin,
-                            boost::intrusive_ptr<Storage_base> &Rin, const unsigned long long &len,
-                            const std::vector<cytnx_uint64> &shape,
-                            const std::vector<cytnx_uint64> &invmapper_L,
-                            const std::vector<cytnx_uint64> &invmapper_R);
-    void Mul_internal_bti16(boost::intrusive_ptr<Storage_base> &out,
-                            boost::intrusive_ptr<Storage_base> &Lin,
-                            boost::intrusive_ptr<Storage_base> &Rin, const unsigned long long &len,
-                            const std::vector<cytnx_uint64> &shape,
-                            const std::vector<cytnx_uint64> &invmapper_L,
-                            const std::vector<cytnx_uint64> &invmapper_R);
-    void Mul_internal_btb(boost::intrusive_ptr<Storage_base> &out,
-                          boost::intrusive_ptr<Storage_base> &Lin,
-                          boost::intrusive_ptr<Storage_base> &Rin, const unsigned long long &len,
-                          const std::vector<cytnx_uint64> &shape,
-                          const std::vector<cytnx_uint64> &invmapper_L,
-                          const std::vector<cytnx_uint64> &invmapper_R);
-
+    template <typename TLin, typename TRin>
+    inline void MulInternalImpl(boost::intrusive_ptr<Storage_base> &out,
+                                boost::intrusive_ptr<Storage_base> &Lin,
```

Comment:
I know this function is the templatized version of the original function. I just wonder `Lin` and `Rin` can be passed by value of instrusive_ptr instead of be passed by reference of instrusive_ptr, right? Or maybe `const&` is also a better choice than reference.

It is totally fine to not make a decision in this PR.